### ISSUE TITLE
Cannot reproduce GenEval 0.95 result with fewer GPUs — plateaus at 0.9 with reward hacking

### ISSUE THREAD

**Issue by avivwei96** [yifan123/flow_grpo#230]
## Problem

I'm trying to reproduce the `geneval_sd3_fast_nocfg` result (GenEval Score reaching 0.95) using 4 GPUs instead of the  32 GPUs. My training plateaus at ~0.9 and does not improve further, despite matching all optimization dynamics. The generated images also show signs of reward hacking — the model appears to game the GenEval evaluation pipeline rather than genuinely improving compositional generation.

## My Setup

- **GPUs**: 4x NVIDIA B200 (192 GB VRAM each)
- **Config**: Adapted `geneval_sd3_fast_nocfg` for 4 GPUs
- **Training script**: `train_sd3_fast.py`

## Config Comparison

| Parameter | Paper (32 GPU) | Mine (4 GPU) |
|---|---|---|
| `train_batch_size` | 9 | 12 |
| `num_image_per_prompt` | 24 | 24 |
| `num_batches_per_epoch` | 4 | 24 |
| `gradient_accumulation_steps` | 2 | 12 |
| **Total samples/epoch** | **1152** | **1152** |
| **group_number** | **48** | **48** |
| **Optimizer steps/epoch** | **2** | **2** |
| **Effective batch/step** | **576** | **576** |
| `learning_rate` | 3e-4 | 3e-4 |
| `clip_range` | 1e-5 | 1e-5 |
| `beta` | 0 | 0 |
| `noise_level` | 0.8 | 0.8 |
| `sde_type` | cps | cps |
| `sde_window_size` | 3 | 3 |
| `same_latent` | False | False |
| `mixed_precision` | fp16 | fp16 |

All other parameters (learning rate, Adam betas, max_grad_norm, etc.) are inherited from `base.py` defaults — identical in both cases.

<details>
<summary>Full 4-GPU config</summary>

```python
def geneval_sd3_fast_nocfg_4gpu():
    gpu_number = 4
    group_number = 48
    config = base.get_config()
    config.dataset = os.path.join(os.getcwd(), "dataset/geneval")
    config.pretrained.model = "stabilityai/stable-diffusion-3.5-medium"
    config.sample.num_steps = 10
    config.sample.eval_num_steps = 40
    config.sample.guidance_scale = 1
    config.sample.eval_guidance_scale = 1
    config.train.cfg = False
    config.resolution = 512
    config.sample.train_batch_size = 12
    config.sample.num_image_per_prompt = 24
    config.sample.num_batches_per_epoch = int(group_number / (gpu_number * config.sample.train_batch_size / config.sample.num_image_per_prompt))
    config.sample.test_batch_size = 14
    config.train.batch_size = config.sample.train_batch_size
    config.train.gradient_accumulation_steps = config.sample.num_batches_per_epoch // 2
    config.train.num_inner_epochs = 1
    config.train.clip_range = 1e-5
    config.train.beta = 0
    config.sample.global_std = True
    config.sample.noise_level = 0.8
    config.sample.sde_window_size = 3
    config.sample.sde_window_range = (0, config.sample.num_steps // 2)
    config.sample.sde_type = "cps"
    config.train.ema = True
    config.use_lora = True
    config.save_freq = 60
    config.eval_freq = 60
    config.reward_fn = {"geneval": 1.0}
    config.prompt_fn = "geneval"
    config.per_prompt_stat_tracking = True
    return config
```
</details>

## Results

- `eval_reward_geneval` reaches ~0.9 by step 800 and plateaus through step 1200+, while the paper reports continuous improvement to 0.95.
- Early training matches well (both reach ~0.78 by the first eval point), but mine stops improving while the paper's continues.
- Generated images look reward-hacked — the paper claims "little to no reward hacking occurred," but I observe visual degradation despite high reward scores.

<img width="3729" height="1610" alt="Image" src="https://github.com/user-attachments/assets/feccf4c8-6253-413f-8c3c-7e8a4e401375" />

<img width="3644" height="1221" alt="Image" src="https://github.com/user-attachments/assets/ff4ee088-3119-460b-a279-bf178b60831a" />

## Questions

1. Are there any hyperparameters or settings not shown in `geneval_sd3_fast_nocfg` that were used for the published 0.95 result?
2. Did you observe any reward hacking during training? If so, how was it mitigated?
3. How many total training steps/epochs were used for the published result?
4. Is there any known issue with reproducing results on fewer GPUs (e.g., micro-batch size 48 vs 288 during gradient accumulation)?
5. Would you recommend adding KL regularization (`beta > 0`) or GRPO-Guard for reproduction on fewer GPUs?

**Comment by lilyandluc** [comment]
how do you know that this config train_sd3_fast.py is for the reported result in the paper?
